drake/systems/frames/CoordinateFrame.h

```diff
@@ -0,0 +1,134 @@
+      os << "  " << c << std::endl;
+    }
+    return os;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const CoordinateFrame& m)
+  {
+    return m.print(os);
+  }
+
+  virtual DrakeSystemPtr setupLCMInputs(const DrakeSystemPtr& sys) { return sys; }
+  virtual DrakeSystemPtr setupLCMOutputs(const DrakeSystemPtr& sys) { return sys; }
+
+  std::string name;  // a descriptive name for this coordinate frame
+
+protected:
+  std::vector<std::string> coordinates; // a string name for each element in the vector (size==dim)
+};
+
+typedef std::shared_ptr<CoordinateFrame> CoordinateFramePtr;
+
+class DLLEXPORT MultiCoordinateFrame : public CoordinateFrame {
+public:
+  MultiCoordinateFrame(const std::string& name, std::initializer_list<std::shared_ptr<CoordinateFrame>> _frames);
+  virtual ~MultiCoordinateFrame(void) {};
+
+  virtual std::ostream& print(std::ostream& os) const {
+    os << "Multi-Coordinate Frame: " << name << " (" << getDim() << " elements)" << std::endl;
+
+    for (auto sf : frames) {
+      os << "  " << sf.frame->name << " (" << sf.frame->getDim() << " elements) " << std::endl;
+    }
+    return os;
+  }
+
+  virtual unsigned int getDim() const { return coordinate_refs.size(); };
+  virtual const std::string& getCoordinateName(unsigned int i) const {
+    if (i>=coordinate_refs.size())
+      throw std::runtime_error("index exceeds dimension of coordinate frame");
+    return coordinate_refs[i].pframe->getCoordinateName(coordinate_refs[i].index_in_subframe);
+  }
+  virtual std::vector<std::string> getCoordinateNames() const {
+    std::vector<std::string> coordinates;
+    for (auto c : coordinate_refs) {
+      coordinates.push_back(c.pframe->getCoordinateName(c.index_in_subframe));
+    }
+    return coordinates;
+  }
+  virtual void setCoordinateNames(const std::vector<std::string>& _coordinates)
+  {
+    throw std::runtime_error("cannot change the names of a multi-coordinate frame directory.  change the subframes.");
+    // note: i could potentially allow this if the sizes were correct, but it seems unlikely to be used for good
+  }
+
+  virtual DrakeSystemPtr setupLCMInputs(const DrakeSystemPtr& sys) { throw std::runtime_error("not implemented yet (need CoordinateTransforms)"); }
+  virtual DrakeSystemPtr setupLCMOutputs(const DrakeSystemPtr& sys) { throw std::runtime_error("not implemented yet (need CoordinateTransforms)"); }
+
+private:
+  /// Store the data in two (compatible) ways:
+  ///   1) as a list of subframes, with a map to the relevant indices in the multi-frame
+  ///   2) as a list of coordinates, with references to the associated subframe
+  struct SubFrame {
+    std::shared_ptr<CoordinateFrame> frame;
+    std::vector<unsigned int> coordinate_indices; // which indices in the multi-frame are associated with this sub-frame
+  };
+  struct CoordinateRef {
+    CoordinateRef(const std::shared_ptr<CoordinateFrame>& _frame) : pframe(_frame.get()) {};
+    CoordinateFrame* pframe;
```

Comment:
on the contrary, i had a shared pointer initially and realized my mistake.  i think it’s really easy to get trigger happy with shared pointers, but we want to avoid assigning priveledges in places where things should not own the pointer.

for this one, to avoid the danger you suggest, I can make this point to the subframe instead.  that’s cleaner.  

i’ve also been changing all the coordinateframe pointers to const pointers.  it’s infectious in the code, but in an awesome cleansing way.

> On Aug 22, 2015, at 9:41 PM, Twan Koolen <notifications@github.com> wrote:
> 
> In drake/systems/frames/CoordinateFrame.h <https://github.com/RobotLocomotion/drake/pull/1274#discussion_r37701727>:
> 
> > +  }
> > +
> > +  virtual DrakeSystemPtr setupLCMInputs(const DrakeSystemPtr& sys) { throw std::runtime_error("not implemented yet (need CoordinateTransforms)"); }
> > +  virtual DrakeSystemPtr setupLCMOutputs(const DrakeSystemPtr& sys) { throw std::runtime_error("not implemented yet (need CoordinateTransforms)"); }
> > +
> > +private:
> > +  /// Store the data in two (compatible) ways:
> > +  ///   1) as a list of subframes, with a map to the relevant indices in the multi-frame
> > +  ///   2) as a list of coordinates, with references to the associated subframe
> > +  struct SubFrame {
> > +    std::shared_ptr<CoordinateFrame> frame;
> > +    std::vector<unsigned int> coordinate_indices; // which indices in the multi-frame are associated with this sub-frame
> > +  };
> > +  struct CoordinateRef {
> > +    CoordinateRef(const std::shared_ptr<CoordinateFrame>& _frame) : pframe(_frame.get()) {};
> > +    CoordinateFrame* pframe;
> why use a raw pointer here? That seems very dangerous when it's initialized from a shared_ptr. Enough things go out of scope in some far away part of the code, and you end up with a dangling pointer.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1274/files#r37701727>.
>